src/backend/distributed/executor/local_executor.c

```diff
@@ -142,32 +143,62 @@ ExecuteLocalTaskList(CitusScanState *scanState, List *taskList)
 	foreach(taskCell, taskList)
 	{
 		Task *task = (Task *) lfirst(taskCell);
+		char *shardQueryString = NULL;
 
-		const char *shardQueryString = TaskQueryString(task);
-		Query *shardQuery = ParseQueryString(shardQueryString, parameterTypes, numParams);
+		PlannedStmt *localPlan = GetCachedLocalPlan(task, distributedPlan);
 
 		/*
-		 * We should not consider using CURSOR_OPT_FORCE_DISTRIBUTED in case of
-		 * intermediate results in the query. That'd trigger ExecuteLocalTaskPlan()
-		 * go through the distributed executor, which we do not want since the
-		 * query is already known to be local.
+		 * If the plan is already cached, don't need to re-plan, just
+		 * acquire necessary locks.
 		 */
-		int cursorOptions = 0;
+		if (localPlan != NULL)
+		{
+			shardQueryString = task->queryStringLazy
+							   ? task->queryStringLazy
+							   : "<optimized out by local execution>";
```

Comment:
I think it's nicer to put this after LogLocalCommand. That way construct the querystring anyway for logging, you have it here.

Reply:
I though it'd be good to pass it to `ExecuteLocalTaskPlan ` as well, no?

Reply:
Yes, I meant move this there. It's not used anywhere else. And it will work for the other branch too, since you call TaskQueryString there (which sets task->queryStringLazy).